src/RA_Defs.cpp

```diff
@@ -6,57 +6,71 @@
 #include <codecvt>
 
 GetParseErrorFunc GetJSONParseErrorStr = GetParseError_En;
+namespace ra {
 
-static_assert(sizeof(BYTE*) == sizeof(char*), "dangerous cast ahead");
-char* DataStreamAsString(DataStream& stream)
-{
-	return reinterpret_cast<char*>(stream.data());
-}
+inline namespace conversions{
+std::string DataStreamAsString(const DataStream& stream) {
 
-std::string Narrow(const wchar_t* wstr)
-{
-	static std::wstring_convert< std::codecvt_utf8_utf16< wchar_t >, wchar_t > converter;
-	return converter.to_bytes(wstr);
+    auto str{ convert_string<std::string>(stream) };
+
+    // pesky null character
+    if (!str.empty()) {
+        str.pop_back();
+    }
+
+    return str; // ok it's definitly showing how it should be
 }
 
-std::string Narrow(const std::wstring& wstr)
-{
-	static std::wstring_convert< std::codecvt_utf8_utf16< wchar_t >, wchar_t > converter;
-	return converter.to_bytes(wstr);
+
+std::string Narrow(const std::wstring& wstr) {
+    return convert_string<std::string>(wstr);
 }
 
-std::wstring Widen(const char* str)
+// N.B.: Not totally sure if you want the strings erased at conversion or not, if so we should rvalues
+// instead.
+std::string Narrow(const wchar_t* wstr)
 {
-	static std::wstring_convert< std::codecvt_utf8_utf16< wchar_t >, wchar_t > converter;
-	return converter.from_bytes(str);
+    std::wstring swstr{ wstr };
+    return convert_string<std::string>(swstr);
 }
 
 std::wstring Widen(const std::string& str)
 {
-	static std::wstring_convert< std::codecvt_utf8_utf16< wchar_t >, wchar_t > converter;
-	return converter.from_bytes(str);
+    return convert_string<std::wstring>(str);
 }
 
-std::wstring Widen(const wchar_t* wstr)
+std::wstring Widen(const char* str)
 {
-	return std::wstring(wstr);
+    std::string std_str{ str };
+    return convert_string<std::wstring>(std_str);
 }
 
-std::wstring Widen(const std::wstring& wstr)
+std::string ByteAddressToString(ByteAddress nAddr)
 {
-	return wstr;
-}
+    // with tinyformat it's easier but I can do it without it
+    // if we used tinyformat it would be like this
+    //return tfm::format("0x%06x", nAddr);
+    // Yup, it's that simple
 
-std::string Narrow(const char* str)
-{
-	return std::string(str);
-}
 
-std::string Narrow(const std::string& str)
-{
-	return str;
+
+    std::ostringstream oss;
+    oss.flags(std::ios::hex | std::ios::basefield | std::ios::showbase);
+    oss << nAddr;
+    return oss.str();
```

Comment:
Will this put leading zeros in front of the value? i.e. `1` => `0x000001`?

Reply:
yes

Reply:
https://youtu.be/uT9Gz27NA3E